### PR TITLE
Made progress widget get values from respective program

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
Fixes https://github.com/mitodl/micromasters/issues/941

#### What's this PR do?
- Calculates total number of courses and passed courses and update progress widget.

#### Where should the reviewer start?
- ProgressWidget.js

#### How should this be manually tested?
- Open ``/dashboard/`` and switch programs

@roberthouse54 @aliceriot @pdpinch 
#### Screenshots (if appropriate)
screen shot 2016-09-20 at 7 05 02 pm

### CODE REVIEW

**Comment by amir-qayyum-khan** [comment]
@roberthouse54 where can I get circle image to match mockups

#### Current version:
screen shot 2016-09-20 at 7 05 02 pm

#### Mockups:
screen shot 2016-08-22 at 3 28 27 pm

cc @pdpinch

**Comment by amir-qayyum-khan** [comment]
@aliceriot or @gsidebo please review this.

File: static/js/util/util.js [hunk 1]
```diff
@@ -365,3 +370,23 @@ export function createForm(url: string, payload: CheckoutPayload): HTMLFormEleme
...
+ */
+export function programCourseInfo(program: Program): Object {
+  let totalCourses = 0;
+  let totalPassedCourses = 0;
+  let passedCourses: Course;
+
+  if (program.courses) {
+    totalCourses = program.courses.length;
+    passedCourses = program.courses.find(course => course.status === STATUS_PASSED);
+    totalPassedCourses = passedCourses ? passedCourses.length : 0;
```
Comment (noisecapella) [inline static/js/util/util.js]: `course.status` was removed recently, there is only status at the run level now. You will need to check  `course.runs[0].status` instead

File: static/js/util/util.js [hunk 2]
```diff
@@ -365,3 +370,23 @@ export function createForm(url: string, payload: CheckoutPayload): HTMLFormEleme
...
+ * Returns total courses and passed courses in program.
+ */
+export function programCourseInfo(program: Program): Object {
+  let totalCourses = 0;
+  let totalPassedCourses = 0;
+  let passedCourses: Course;
+
+  if (program.courses) {
+    totalCourses = program.courses.length;
+    passedCourses = program.courses.find(course => course.status === STATUS_PASSED);
```
Comment (noisecapella) [inline static/js/util/util.js]: `find` will only get one course. Did you mean to use `filter`?

**Comment by amir-qayyum-khan** [comment]
fixed @noisecapella

File: static/js/components/ProgressWidget_test.js [hunk 3]
```diff
@@ -32,7 +32,9 @@ describe('ProgressWidget', () => {
     ));
     const wrapper = shallow(<ProgressWidget program={program}/>);
     const totalCourses = program.courses.length;
-    const passedCourses = program.courses.find(course => course.status === STATUS_PASSED);
+    const passedCourses = program.courses.filter(
+      course => course.runs.length > 0 && course.runs[0].status === STATUS_PASSED
+    );
     const totalPassedCourses = passedCourses ? passedCourses.length : 0;
```
Comment (noisecapella) [inline static/js/components/ProgressWidget_test.js]: There shouldn't be any case where `passedCourses` is not a list, so we don't need the trinary expression here

File: static/js/util/util.js [hunk 4]
```diff
@@ -381,7 +381,9 @@ export function programCourseInfo(program: Program): Object {
 
   if (program.courses) {
     totalCourses = program.courses.length;
-    passedCourses = program.courses.find(course => course.status === STATUS_PASSED);
+    passedCourses = program.courses.filter(
+      course => course.runs.length > 0 && course.runs[0].status === STATUS_PASSED
+    );
     totalPassedCourses = passedCourses ? passedCourses.length : 0;
```
Comment (noisecapella) [inline static/js/util/util.js]: Just use `passedCourses.length` here too

File: static/js/util/util_test.js [hunk 5]
```diff
@@ -421,7 +421,9 @@ describe('utility functions', () => {
       ));
       const programInfoActual = programCourseInfo(program);
       const totalCoursesExpected = program.courses.length;
-      const passedCourses = program.courses.find(course => course.status === STATUS_PASSED);
+      const passedCourses = program.courses.filter(
+        course => course.runs.length > 0 && course.runs[0].status === STATUS_PASSED
+      );
       const totalPassedCoursesExpected = passedCourses ? passedCourses.length : 0;
```
Comment (noisecapella) [inline static/js/util/util_test.js]: Use `passedCourses.length` here too

**Comment by amir-qayyum-khan** [comment]
Done @noisecapella

File: static/js/components/ProgressWidget.js [hunk 6]
```diff
@@ -3,30 +3,28 @@ import React from 'react';
...
+    totalPassedCourses: number,
+    totalCourses: number
   ): React$Element<*> => {
     const radiusForMeasures = radius - strokeWidth / 2;
     const width = radius * 2;
     const height = radius * 2;
     const viewBox = `0 0 ${width} ${height}`;
     const dashArray = radiusForMeasures * Math.PI * 2;
-    const dashOffset = dashArray - dashArray * actual / total;
+    const dashOffset = dashArray - dashArray * totalPassedCourses / totalCourses;
```
Comment (noisecapella) [inline static/js/components/ProgressWidget.js]: Can you change this to handle a division by zero here?

Comment (amir-qayyum-khan) [inline static/js/components/ProgressWidget.js]: Fixed @noisecapella

**Comment by noisecapella** [comment]
Looks great :+1:

File: static/js/util/util_test.js [hunk 7]
```diff
@@ -409,4 +412,24 @@ describe('utility functions', () => {
...
+      let program = findProgram(course => (
+        course.runs.length > 0 &&
+        course.runs[0].status === STATUS_PASSED
+      ));
+      const programInfoActual = programCourseInfo(program);
+      const totalCoursesExpected = program.courses.length;
+      const passedCourses = program.courses.filter(
+        course => course.runs.length > 0 && course.runs[0].status === STATUS_PASSED
+      );
+      const totalPassedCoursesExpected = passedCourses.length;
```
Comment (gsidebo) [inline static/js/util/util_test.js]: i made a similar comment on a previous PR, but the approach of this test is problematic. unit tests like these should be performed with exact, understood data, and should test that the outcomes based on that data match our expectations. what you're doing here is fetching a program that matches a couple conditions, but really we don't know what the data will look like exactly. after that, you're directly copying the body of the tested function and testing your expectations against that. 

instead, you should create a program with exact data specifically for this test, pass it into the function, and make sure that the values we get back are what we expect. if you want to create the program based off of an existing one in our `DASHBOARD_RESPONSE` variable, that might work too

Comment (amir-qayyum-khan) [inline static/js/util/util_test.js]: fixed @gsidebo

**Comment by gsidebo** [comment]
looks great 👍  thanks!

**Pull request merged**